Sources/ServiceLifecycle/ServiceRunner.swift

```diff
@@ -0,0 +1,383 @@
+//
+// This source file is part of the SwiftServiceLifecycle open source project
+//
+// Copyright (c) 2023 Apple Inc. and the SwiftServiceLifecycle project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of SwiftServiceLifecycle project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+import Logging
+import UnixSignals
+
+/// A ``ServiceRunner`` is responsible for running a number of services, setting up signal handling and signalling graceful shutdown to the services.
+public actor ServiceRunner: Sendable {
+    /// The internal state of the ``ServiceRunner``.
+    private enum State {
+        /// The initial state of the runner.
+        case initial
+        /// The state once ``ServiceRunner/run()`` has been called.
+        case running(
+            gracefulShutdownStreamContinuation: AsyncStream<Void>.Continuation
+        )
+        /// The state once ``ServiceRunner/run()`` has finished.
+        case finished
+    }
+
+    /// The services to run.
+    private let services: [any Service]
+    /// The runner's configuration.
+    private let configuration: ServiceRunnerConfiguration
+    /// The logger.
+    private let logger: Logger
+
+    /// The current state of the runner.
+    private var state: State = .initial
+
+    /// Initializes a new ``ServiceRunner``.
+    ///
+    /// - Parameters:
+    ///   - services: The services to run.
+    ///   - configuration: The runner's configuration.
+    ///   - logger: The logger.
+    public init(
+        services: [any Service],
+        configuration: ServiceRunnerConfiguration,
+        logger: Logger
+    ) {
+        self.services = services
+        self.configuration = configuration
+        self.logger = logger
+    }
+
+    /// Runs all the services by spinning up a child task per service.
+    /// Furthermore, this method sets up the correct signal handlers
+    /// for graceful shutdown.
+    public func run(file: String = #file, line: Int = #line) async throws {
+        switch self.state {
+        case .initial:
+            guard !self.services.isEmpty else {
+                self.state = .finished
+                return
+            }
+
+            let (gracefulShutdownStream, gracefulShutdownContinuation) = AsyncStream.makeStream(of: Void.self)
+
+            self.state = .running(
+                gracefulShutdownStreamContinuation: gracefulShutdownContinuation
+            )
+
+            var potentialError: Error?
+            do {
+                try await self._run(gracefulShutdownStream: gracefulShutdownStream)
+            } catch {
+                potentialError = error
+            }
+
+            switch self.state {
+            case .initial, .finished:
+                fatalError("ServiceRunner is in an invalid state \(self.state)")
```

Comment:
unless there is a memory corruption / security risk, i am not a fan of fatalError

Reply:
but this is a clear programer error isn't it?

Reply:
Exactly what @fabianfett said here. This is a programmer error on our side and we are using `fatalError`s for these quite consistently across NIO. In this particular instance if we are in any other state than `running` we made a mistake in our implementation.